cpp/src/arrow/compute/kernels/compare.cc

```diff
@@ -0,0 +1,175 @@
+#include "arrow/compute/kernels/filter.h"
+#include "arrow/compute/kernels/util-internal.h"
+#include "arrow/util/bit-util.h"
+#include "arrow/util/logging.h"
+
+namespace arrow {
+
+namespace compute {
+
+class FunctionContext;
+struct Datum;
+
+template <typename ArrowType, CompareOperator Op,
+          typename ArrayType = typename TypeTraits<ArrowType>::ArrayType,
+          typename ScalarType = typename TypeTraits<ArrowType>::ScalarType,
+          typename T = typename TypeTraits<ArrowType>::CType>
+static Status CompareArrayScalar(const ArrayData& input, const ScalarType& scalar,
+                                 uint8_t* bitmap) {
+  const T right = scalar.value;
+  const T* values = input.GetValues<T>(1);
+
+  size_t i = 0;
+  internal::GenerateBitsUnrolled(bitmap, 0, input.length, [values, right, &i]() -> bool {
+    return Comparator<T, Op>::Compare(values[i++], right);
+  });
+
+  return Status::OK();
+}
+
+template <typename ArrowType, CompareOperator Op>
+class CompareFunction final : public FilterFunction {
+  using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+  using ScalarType = typename TypeTraits<ArrowType>::ScalarType;
+
+ public:
+  explicit CompareFunction(FunctionContext* ctx) : ctx_(ctx) {}
+
+  Status Filter(const ArrayData& input, const Scalar& scalar, ArrayData* output) const {
+    // Caller must cast
+    DCHECK(input.type->Equals(scalar.type));
+    // Output must be a boolean array
+    DCHECK(output->type->Equals(boolean()));
+    // Output must be of same length
+    DCHECK_EQ(output->length, input.length);
+
+    // Scalar is null, all comparisons are null.
+    if (!scalar.is_valid) {
+      return detail::SetAllNulls(ctx_, input, output);
+    }
+
+    // Copy null_bitmap
+    RETURN_NOT_OK(detail::PropagateNulls(ctx_, input, output));
+
+    uint8_t* bitmap_result = output->buffers[1]->mutable_data();
+    return CompareArrayScalar<ArrowType, Op>(
+        input, static_cast<const ScalarType&>(scalar), bitmap_result);
+  }
+
+ private:
+  FunctionContext* ctx_;
+};
+
+template <typename ArrowType, CompareOperator Op>
+static inline std::shared_ptr<FilterFunction> MakeCompareFunctionTypeOp(
+    FunctionContext* ctx) {
+  return std::make_shared<CompareFunction<ArrowType, Op>>(ctx);
+}
+
+template <typename ArrowType>
+static inline std::shared_ptr<FilterFunction> MakeCompareFilterFunctionType(
+    FunctionContext* ctx, struct CompareOptions options) {
+  switch (options.op) {
+    case CompareOperator::EQUAL:
+      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::EQUAL>(ctx);
+    case CompareOperator::NOT_EQUAL:
+      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::NOT_EQUAL>(ctx);
+    case CompareOperator::GREATER:
+      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::GREATER>(ctx);
+    case CompareOperator::GREATER_EQUAL:
+      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::GREATER_EQUAL>(ctx);
+    case CompareOperator::LOWER:
```

Comment:
@fsaintjacques I think `LESS` is more natural than `LOWER` for the opposite word for `GREATER` because I've not much seen the use of `lower` and `greater` pair.  For `lower`, `higher` is more natural for me (but I'm not good for English, so this is just my sense).

Moreover, STL provides `less` and `greater` pair, so I think `less` is natural at least in C++ API.

Reply:
WIll do https://jira.apache.org/jira/browse/ARROW-4774

Reply:
Maybe you attempted to paste https://jira.apache.org/jira/browse/ARROW-5171
Thank you very much 😄